server/index.js

```diff
@@ -37,8 +37,8 @@ async function main() {
    * operate very rapidly to enable quick as possible server startup.
    */
   await middleware.installDatabasePools(app);
-  await middleware.installSession(app);
   await middleware.installPassport(app);
+  await middleware.installSession(app);
```

Comment:
I think sessions have to be installed before Passport?

Reply:
@plitzenberger perhaps this is your passport.initialize issue?

Reply:
I've tried it actually with both versions. Same issue.

Reply:
But I'll definitely check this again. Thx for your help @benjie :)

Reply:
Let me know; I can pull it down and try myself at some point, but I’d expect it to work.

Reply:
Just checked again. The order doesn't seem to have any effect. I'm going to have a deeper look.

Reply:
so I was just debugging to see why the github authentication doesn't work when I have redis enabled and there I saw the same line where it fails to work with login/password:
```
// node_modules/passport/lib/http/request.js:46
if (!req._passport) { return this.error(new Error('passport.initialize() middleware not in use')); }
```

and now `req._passport` was not `undefined`, like when I used your code to login with username/password. 

disclaimer: in my `index.js` `installSession()` comes before `installPassport()` because I just read it in the passportjs docs. and I replaced redis with `express-session`,

here is my callstack: 
```
// not working (username/passport strategy)
req.login.req.logIn (node_modules/passport/lib/http/request.js:46)
login (e:\__Work\development\apps\_forks\bootstrap-react-apollo\server\plugins\authentication.js:113)
processTicksAndRejections (task_queues.js:89)
[ async function ]
login (e:\__Work\development\apps\_forks\bootstrap-react-apollo\server\plugins\authentication.js:103)
newResolver (makeExtendSchemaPlugin.js:304)
resolveFieldValueOrError (execute.js:480)
resolveField (execute.js:447)
(anonymous function) (execute.js:263)
(anonymous function) (promiseReduce.js:32)
promiseReduce (promiseReduce.js:29)
executeFieldsSerially (execute.js:260)
executeOperation (execute.js:238)
executeImpl (execute.js:85)
execute (execute.js:62)
(anonymous function) (createPostGraphileHttpRequestHandler.js:541)
(anonymous function) (createPostGraphileHttpRequestHandler.js:83)
(anonymous function) (withPostGraphileContext.js:145)
withAuthenticatedPgClient (withPostGraphileContext.js:89)
processTicksAndRejections (task_queues.js:89)

```

what I saw immediatly as wrong is that `this` in the working (github) version is an instance of `IncomingMessage` and in the not working version is `global`

so I assume maybe it has to be initialized the same way like `installPassportStrategy.js` does it?

Reply:
Ok, I found it, it seems `login` needs to called from `req` so `this` in `login` scope is `IncomingMessage`.

This is the important change:
```
// server/plugins/authentication.js:113
 await req.login(user, () => { console.debug(`${user.username} is logged in`) });
```

Here is my diff:
https://github.com/graphile/bootstrap-react-apollo/pull/11/commits/e802dc4dd893526ce50768691385a3adece10995

BUT I'm not really sure if it's alright to but the whole `req` inside `context`...? @benjie

Reply:
No, instead bind it, e.g. something like `login: req.login.bind(req)` (am on phone, so this might not relate to the actual code 🤷‍♂️)

Reply:
Thx guys. Binding to req and adding a callback to the login function did the trick. Now it's working like expected 🥳

Reply:
great! seems I need to brush up on my scope binding game though ;0